### PR TITLE
Scalarize the vector inputs to llvm.lround intrinsic by default.

### PR DESCRIPTION
**Pull request by sgundapa**
Verifier is updated in a different patch to let the vector types for llvm.lround and llvm.llround intrinsics.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/lround.ll [hunk 1]
```diff
@@ -1034,3 +1037,507 @@ entry:
...
+; GFX11-GISEL-NEXT:    v_cmp_ge_f32_e64 s0, |v5|, 0.5
+; GFX11-GISEL-NEXT:    s_delay_alu instid0(VALU_DEP_2) | instskip(NEXT) | instid1(VALU_DEP_2)
+; GFX11-GISEL-NEXT:    v_and_or_b32 v0, 0x80000000, v0, v4
+; GFX11-GISEL-NEXT:    v_cndmask_b32_e64 v5, 0, 1.0, s0
+; GFX11-GISEL-NEXT:    s_delay_alu instid0(VALU_DEP_1) | instskip(NEXT) | instid1(VALU_DEP_1)
+; GFX11-GISEL-NEXT:    v_and_or_b32 v1, 0x80000000, v1, v5
+; GFX11-GISEL-NEXT:    v_dual_add_f32 v0, v2, v0 :: v_dual_add_f32 v1, v3, v1
+; GFX11-GISEL-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %0 = tail call <2 x float> @llvm.round.v2f32.v2f32(<2 x float> %arg)
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/lround.ll:1149]: These regular round cases are probably already tested, and shouldn't be in the "lround" test

File: llvm/test/CodeGen/AMDGPU/lround.ll [hunk 2]
```diff
@@ -14,6 +14,9 @@ declare i64 @llvm.lround.i64.f64(double)
 declare i64 @llvm.llround.i64.f32(float)
 declare half @llvm.round.f16(half)
 declare i32 @llvm.lround.i32.f16(half %arg)
+declare <2 x float> @llvm.round.v2f32.v2f32(<2 x float> %arg)
+declare <2 x i32> @llvm.lround.v2i32.v2f32(<2 x float> %arg)
+declare <2 x i64> @llvm.lround.v2i64.v2f32(<2 x float> %arg)
```
Comment (tschuett) [inline llvm/test/CodeGen/AMDGPU/lround.ll:19]: The `declare` s are a waste of your time:
https://discourse.llvm.org/t/recent-improvements-to-the-ir-parser/77366

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 3]
```diff
@@ -4920,6 +4920,8 @@ LegalizerHelper::fewerElementsVector(MachineInstr &MI, unsigned TypeIdx,
   case G_INTRINSIC_LLRINT:
   case G_INTRINSIC_ROUND:
   case G_INTRINSIC_ROUNDEVEN:
+  case G_LROUND:
+  case G_LLROUND:
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:4924]: unrelated changes?

Comment (sgundapa) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:4924]: The lit test is not testing this but it will complete the handling of cases in line with similar intrinsic, LRINT.

File: llvm/test/CodeGen/AMDGPU/lround.ll (hunk 2 above)
Comment (sgundapa) [reply llvm/test/CodeGen/AMDGPU/lround.ll:19]: I will remove the lines

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 4]
```diff
@@ -2063,6 +2063,12 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
     case Intrinsic::roundeven:
       ISD = ISD::FROUNDEVEN;
       break;
+    case Intrinsic::lround:
+      ISD = ISD::LROUND;
+      break;
+    case Intrinsic::llround:
+      ISD = ISD::LLROUND;
+      break;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:2071]: This is a separate change that should get cost model tests

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 5]
```diff
@@ -17856,15 +17858,17 @@ SDValue DAGCombiner::visitFP_TO_UINT(SDNode *N) {
   return FoldIntToFPToInt(N, DAG);
 }
 
-SDValue DAGCombiner::visitXRINT(SDNode *N) {
+SDValue DAGCombiner::visitXROUND(SDNode *N) {
   SDValue N0 = N->getOperand(0);
   EVT VT = N->getValueType(0);
 
   // fold (lrint|llrint undef) -> undef
+  // fold (lround|llround undef) -> undef
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:17866]: I don't think this is tested. This *might* happen for the v3f16 case

**Review by arsenm — APPROVED** [review]

**Comment by sgundapa** [comment]
I will rebase for a clean build and merge this change

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h (hunk 4 above)
Comment (sgundapa) [reply llvm/include/llvm/CodeGen/BasicTTIImpl.h]: pushed a patch - https://github.com/llvm/llvm-project/pull/105694
